sbi/inference/npse/npse.py

```diff
@@ -0,0 +1,612 @@
+
+
+class NPSE(NeuralInference):
+    def __init__(
+        self,
+        prior: Optional[Distribution] = None,
+        score_estimator: Union[str, Callable] = "mlp",
+        sde_type: str = "ve",
+        device: str = "cpu",
+        logging_level: Union[int, str] = "WARNING",
+        summary_writer: Optional[SummaryWriter] = None,
+        show_progress_bars: bool = True,
+        **kwargs,
+    ):
+        """Base class for Neural Posterior Score Estimation methods.
+
+        Instead of performing conditonal *density* estimation, NPSE methods perform
+        conditional *score* estimation i.e. they estimate the gradient of the log
+        density using denoising score matching loss. We not only estimate the score
+        of the posterior, but a family of distributions analogous to diffusion models.
+
+        NOTE: Single-round NPSE is currently the only supported mode.
+
+        Args:
+            prior: Prior distribution.
+            score_estimator: Neural network architecture for the score estimator. Can be
+                a string (e.g. 'mlp') or a callable that returns a neural network.
+            sde_type: Type of SDE to use. Must be one of ['vp', 've', 'subvp'].
+            device: Device to run the training on.
+            logging_level: Logging level for the training. Can be an integer or a
+                string.
+            summary_writer: Tensorboard summary writer.
+            show_progress_bars: Whether to show progress bars during training.
+            kwargs: Additional keyword arguments.
+
+        References:
+            - Geffner, Tomas, George Papamakarios, and Andriy Mnih. "Score modeling for
+              simulation-based inference." NeurIPS 2022 Workshop on Score-Based Methods.
+              2022.
+            - Sharrock, Louis, et al. "Sequential neural score estimation: Likelihood-
+              free inference with conditional score based diffusion models." arXiv
+              preprint arXiv:2210.04872 (2022).
+        """
+
+        super().__init__(
+            prior=prior,
+            device=device,
+            logging_level=logging_level,
+            summary_writer=summary_writer,
+            show_progress_bars=show_progress_bars,
+        )
+
+        # As detailed in the docstring, `score_estimator` is either a string or
+        # a callable. The function creating the neural network is attached to
+        # `_build_neural_net`. It will be called in the first round and receive
+        # thetas and xs as inputs, so that they can be used for shape inference and
+        # potentially for z-scoring.
+        check_estimator_arg(score_estimator)
+        if isinstance(score_estimator, str):
+            self._build_neural_net = posterior_score_nn(
+                sde_type=sde_type, score_net_type=score_estimator, **kwargs
+            )
+        else:
+            self._build_neural_net = score_estimator
+
+        self._proposal_roundwise = []
+
+    def append_simulations(
```

Comment:
Why do we need this and not just `base.append_simulations`? Just for the error message on multi-round?